rust/common-ifc/src/graph/validation.rs

```diff
@@ -0,0 +1,227 @@
+/// to another node in the graph, an input port doesn't have
+/// multiple incoming connections, and all edges with this node are valid.
+fn validate_node_connections<G: PortGraph + ?Sized>(graph: &G, node: &G::Node) -> Result<()> {
+    let id = node.id();
+    let mut has_connections = false;
+    let mut incoming_set = HashSet::new();
+
+    for edge in graph.get_connections(node, None) {
+        has_connections = true;
+        let source = edge.source();
+        let target = edge.target();
+        validate_port_exists::<G>(graph, source.0, source.1, PortType::Output)?;
+        validate_port_exists::<G>(graph, target.0, target.1, PortType::Input)?;
+
+        if target.0 == id && !incoming_set.insert(target.1) {
+            return Err(CommonIfcError::InvalidGraph(format!(
+                "Input '{}' on node '{}' has multiple incoming connections.",
+                target.1, id
+            )));
+        }
+    }
+
+    if !has_connections {
+        Err(CommonIfcError::InvalidGraph(format!(
+            "Node '{}' is not connected to any other nodes.",
+            id
+        )))
+    } else {
+        Ok(())
+    }
+}
+
+/// Validates the existence of a node with given `node_id`
+/// in the hash map with `port_type` port named `port_name`.
+fn validate_port_exists<G: PortGraph + ?Sized>(
+    graph: &G,
+    node_id: &G::NodeId,
+    port_name: &G::PortName,
+    port_type: PortType,
+) -> Result<()> {
+    if let Some(node) = graph.get_node(node_id) {
+        if match port_type {
+            PortType::Input => node.inputs().any(|p| p == port_name),
+            PortType::Output => node.outputs().any(|p| p == port_name),
+        } {
+            return Ok(());
+        }
+    } else {
+        return Err(CommonIfcError::InvalidGraph(format!(
+            "Node '{}' is not in the graph.",
+            node_id,
+        )));
+    }
+
+    let port_type_name = match port_type {
+        PortType::Input => "input",
+        PortType::Output => "output",
+    };
+    Err(CommonIfcError::InvalidGraph(format!(
+        "Node '{}' does not contain an {} port with the name '{}'.",
+        node_id, port_type_name, port_name,
+    )))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::PortGraph;
+    use crate::graph::TestGraph;
+
+    #[test]
+    fn it_validates_graph() -> crate::Result<()> {
+        let mut graph = TestGraph::from_iters(
+            [
+                ("A", vec![], vec!["a_out_1", "a_out_2"]),
+                ("B", vec!["b_in"], vec!["b_out"]),
+                ("C", vec!["c_in"], vec!["c_out"]),
+                ("D", vec!["d_in_1", "d_in_2", "d_in_3"], vec![]),
+            ],
+            [
+                (("A", "a_out_1"), ("B", "b_in")),
+                (("A", "a_out_2"), ("C", "c_in")),
+                (("B", "b_out"), ("D", "d_in_1")),
+                (("C", "c_out"), ("D", "d_in_2")),
+            ],
+        );
+        graph.validate_port_graph()?;
+
+        // Unconnected nodes should fail.
```

Comment:
In the future, it would be nice to break these out into discrete tests

Reply:
The rational here was ensuring the expected error was hit (and not a different graph validation failure). Having structured validation failures could make these tests more concise and targeted:

```rust
CommonIfcError::InvalidGraph(GraphValidationError::InvalidConnection { node, port })
```